### PR TITLE
Add custom attributes sources block support

### PR DESCRIPTION
**Pull request by youknowriad**
This PR explores the possibility to allow users and extenders to flag that some block attributes come from external sources. This POC focuses on implementing the "custom fields" external source while providing a very basic API to define alternative external sources for block attributes.

**For instance:** this allows paragraph block to have their content sourced from a post meta.

This project have been renamed "block bindings" and you can read more details about it in #53300

### CODE REVIEW

**Comment by youknowriad** [comment]
There's currently a bug in this PR when you click `Enter` on a custom sourced paragraph block. or when you paste something that is "multi-block" within it.

For instance, now, when clicking "enter", we try to split the block into two, the first block remained externally sourced but it gets a "content" attribute filled as well (bug you can see in the code editor mode) while the second paragraph is not externally sourced.  

Something similar happens when pasting (you might also see two externally sourced blocks one at the beginning and one at the end)

---
I guess we need to figure out how to address this, it feels like the ideal solution would be to not allow splitting and for pasting I'm not sure what the expected behavior should be (maybe paste as text within the block). cc @ellatrix @mtias

**Comment by mtias** [comment]
There's a few layers to this. Eventually, I think the attribute source plus the block's interface should be dictating the "type" for a field, meaning that if we have a coupling of "content" attribute being handled by a `RichText` component as its setter, then the field would be a rich-text field (which we don't really have a concept for today). In the absence of that, it seems we could treat a string-bounded custom field attribute as a locked down rich text component, or switch to `PlainText`. Ideally we don't need to handle these explicitly in the block itself, and it's something that just flows transparently between attribute declaration, attribute setters, and rich text component values.

**Comment by youknowriad** [comment]
This PR served its purpose as an experiment and has been implemented in other PRs.

**Comment by gziolo** [comment]
Thank you for sharing the prototype that helped to clarify further work necessary like:

- https://github.com/WordPress/gutenberg/issues/53300
- https://github.com/WordPress/gutenberg/issues/54536

File: packages/editor/src/attributes-custom-sources/meta.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/**
```
Comment (youknowriad) [inline packages/editor/src/attributes-custom-sources/meta.js:1]: This file is an example of frontend definition of a custom source.

File: lib/custom-sources/meta.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
+ * Meta Custom Source
+ *
+ * @package gutenberg
+ */
+
+return array(
```
Comment (youknowriad) [inline lib/custom-sources/meta.php:8]: And this is the definition of a custom source in the backend.

**Pull request closed without merging**